src/commands/watch.rs

```diff
@@ -145,116 +173,124 @@ impl WatchCommand {
-                                c.id
-                            );
-                            None
-                        }
-                    }
-                } else {
-                    // No build config for this component so lets watch the source instead
-                    if let RawModuleSource::FileReference(path) = &c.source {
-                        if let Some(string_path) = path.to_str() {
-                            return Some(vec![String::from(string_path)]);
-                        }
-                    }
-                    None
-                }
+                let build = c.build.as_ref()?;
+                let Some(watch) = build.watch.clone() else {
+                    eprintln!(
+                        "You haven't configured what to watch for the component: '{}'. Learn how to configure Spin watch at https://developer.fermyon.com/common/cli-reference#watch",
+                        c.id
+                    );
+                    return None;
+                };
+                let sources = build.workdir.clone().map(|workdir| watch.iter()
+                        .filter_map(|w| workdir.join(w).to_str().map(String::from))
+                        .collect()).unwrap_or(watch);
+                Some(match self.skip_build {
+                    false => sources
+                        .into_iter()
+                        .map(|s| WatchPattern::new(s, true, app_dir.as_path()))
+                        .collect::<Vec<Result<WatchPattern>>>(),
+                    true => sources
+                        .into_iter()
+                        .map(|s| WatchPattern::new(s, false, app_dir.as_path()))
+                        .collect(),
+                })
             })
-            .flatten();
-        let component_source_patterns = app_manifest
+            .flatten()
+            .collect::<Result<Vec<WatchPattern>>>()?;
+
+        // We want to watch a component's source if it has no build step. If we're skipping
+        // building then we'll watch all component sources.
+        let mut artifact_patterns = app_manifest
             .components
             .iter()
             .filter_map(|c| {
-                if let RawModuleSource::FileReference(path) = &c.source {
-                    return path.to_str();
-                }
-                None
+                let RawModuleSource::FileReference(path) = &c.source else {
+                    return None;
+                };
+                let Some(path_str) = path.to_str() else {
+                    return None;
+                };
+                let watch_this_source = self.skip_build || c.build.is_none();
+                Some(match watch_this_source {
+                    false => WatchPattern::new(path_str.to_owned(), false, app_dir.as_path()),
+                    true => WatchPattern::new(path_str.to_owned(), true, app_dir.as_path()),
+                })
```

Comment:
I suspect you meant to do:

```suggestion
                Some(WatchPattern::new(path_str.to_owned(), watch_this_source, app_dir.as_path()))
```